Rename hprof to tprof

josevalim opened the pull request:
Instead of having 4 profilers with 4 different APIs, this pull request unifies call, time, and memory profiling under a single module.

The goal is to eventually deprecate and remove cprof and eprof and have only two profiling modules: `fprof` and `tprof`, which are understandably different as they provide different functionalities and measurements.

Here is the new tprof in practice:

```
1> tprof:profile(base64, encode, [crypto:strong_rand_bytes(1024000)]).
FUNCTION                                         CALLS  [    %]
rpc:module_info/1                                    1  [ 0.00]
inet_config:module_info/1                            1  [ 0.00]
io_lib_format:module_info/1                          1  [ 0.00]
kernel_config:module_info/1                          1  [ 0.00]
filename:pathtype/1                                  1  [ 0.00]
filename:unix_pathtype/1                             1  [ 0.00]
filename:module_info/1                               1  [ 0.00]
user_sup:module_info/1                               1  [ 0.00]
erl_scan:module_info/1                               1  [ 0.00]
unicode:module_info/1                                1  [ 0.00]
...ELLIDED...
lists:keyfind/3                                     57  [ 0.28]
erl_prim_loader:archive_split/3                     62  [ 0.31]
erts_internal:trace_pattern/3                       78  [ 0.39]
hprof:collect_trace/4                               83  [ 0.41]
hprof:'-collect/2-fun-0-'/4                         83  [ 0.41]
maps:try_next/2                                    101  [ 0.50]
maps:fold_1/4                                      102  [ 0.51]
hprof:'-enable_pattern/5-fun-0-'/3                 112  [ 0.56]
erlang:trace_info/2                               1024  [ 5.11]
lists:foldl_1/3                                   4301  [21.45]
hprof:combine_trace/1                             6802  [33.92]
hprof:'-collect_trace/4-fun-0-'/4                 6869  [34.25]
                                                        [100.0]
ok
2> tprof:profile(base64, encode, [crypto:strong_rand_bytes(1024000)], #{type => call_time}).

****** Process <0.96.0>  --  100.00% of total ***
FUNCTION                               CALLS  TIME us   PER CALL  [    %]
erts_internal:trace/3                      1        0         0  [ 0.00]
erlang:trace/3                             1        0         0  [ 0.00]
erlang:ensure_tracer_module_loaded/2       1        0         0  [ 0.00]
base64:encode/1                            1        0         0  [ 0.00]
base64:get_padding/1                       1        0         0  [ 0.00]
base64:get_encoding_offset/1               1        0         0  [ 0.00]
lists:keyfind/3                            1        0         0  [ 0.00]
base64:encode/2                            1        1         1  [ 0.01]
base64:encode_binary/4                170668     7509         0  [99.99]
                                                 7510            [100.0]
ok
3> tprof:profile(base64, encode, [crypto:strong_rand_bytes(1024000)], #{type => call_memory}).

****** Process <0.99.0>  --  100.00% of total ***
FUNCTION                 CALLS  WORDS  PER CALL  [    %]
base64:encode_binary/4  170668      5         0  [31.25]
base64:encode/2              1     11        11  [68.75]
                                   16            [100.0]
ok
```

Why use the `hprof` instead of the `eprof` API? [The arguments were outlined by @max-au here](https://github.com/erlang/otp/pull/6639#issuecomment-1374947454).

I have not yet updated docs nor tests. I have some open questions:

1. Do we want to go ahead with this?

2. What is the default measurement? Currently it is `call_count`, as it is the most basic measurement, but it is not quite useful.

3. That said, do we even want to keep `call_count` or `cprof` around? I have never used it in practice, it is extremely verbose and `eprof/hprof` are fast enough to be kept as defaults. So I would suggest nuking `call_count` and `cprof` altogether.

josevalim commented:
There are some important design decisions I have taken:

1. I kept all measurements as `{Pid, Calls, Value}`. For `call_count`, `Calls == Value` and `Pid == all`. When printing `call_count`, we don't include a column for `Value` nor `Value per Call` (as Value == Calls)`.

2. Treating `Pid == all` works elegantly because we can compute totals. Computing a total transforms `#{pid() => [Measurements]}` into `#{all => [Measurements]}`. Which means both call_count (which works across all PIDs) and totaling call_time/call_memory across all PIDs emit the same data structure at the end.

3. I merged module+fa columns on printed results (similar to eprof) as otherwise long module names take too much horizontal space

max-au commented:
I like this! Mainly because it brings a more modern API to `eprof` and `cprof`, it would be much easier to run those against supervision trees.
Such unification makes it very clear that you should not be running multiple (tracing) profilers in the same ERTS (as results may be at the very least surprising).
One thing that might be useful is to publish more blogs/articles on `hprof` (and consequently `eprof`) usage. I wrote a bunch of internal documentation that might still exist, although I no longer have access to it. Probably @potatosalad could rescue some...

josevalim commented:
@max-au @jhogberg I have updated all docs and I have updated the tests. I still need to add tests for the new functionality but, before I do that, I would like to hear your thoughts on two open questions:

1. Should we have `call_count`? `call_count` does not provide a lot of useful information and it always gather information across all processes (so some options such as `set_on_spawn` and `rootset` have no effect). The only reason would be performance, if in a system computing `call_count` is much faster than `call_time`/`call_memory`?

2. Which one should be the default? `call_count`, `call_time` or `call_memory`?

jhogberg commented:
> 1. Should we have `call_count`? `call_count` does not provide a lot of useful information and it always gather information across all processes (so some options such as `set_on_spawn` and `rootset` have no effect). The only reason would be performance, if in a system computing `call_count` is much faster than `call_time`/`call_memory`?

That depends, do you think keeping it complicates things a lot?

> 2. Which one should be the default? call_count, call_time or call_memory?

Perhaps `call_memory`, since most people wanting `call_time` would probably use `perf` these days?

> PS: The docs are failing but I don't know why. If anyone has an idea, let me know. It says "ERROR tprof:inspect/3 unknown spec" but such as spec exists. :)

The doc refers to two different clauses for `tprof:inspect/3`, while the spec has only one clause.

josevalim commented:
> That depends, do you think keeping it complicates things a lot?

In terms of implementation, it does not. A dozen LOC at best. The only downside it does not support the same options as the other two.

However, if we pick a default of `call_memory`, would there be any interest in using `call_count`? I would say only if the performance cost of either `call_time` or `call_memory` is too much. Would that ever be the case?

garazdawi commented:
In my view `call_time` should be default, as `perf` is not something that is available on Windows/MacOS and there are a lot of people developing on such systems. For the same reason we should keep `call_count`  around, so that users can have a lighter weight profiling tool on those OSs.

mikpe commented:
> In my view `call_time` should be default, as `perf` is not something that is available on Windows/MacOS and there are a lot of people developing on such systems. For the same reason we should keep `call_count` around, so that users can have a lighter weight profiling tool on those OSs.

I needed to use cprof recently on a production system since perf would have required a restart with different VM parameters, and that wasn't an option (due to $reasons). Please keep `call_count`.

josevalim commented:
@mikpe for curiosity's sake: could you have used eprof? If not, why not?

mikpe commented:
I used `eprof` too. Then `fprof` but that was delicate -- I had to shorten the run to avoid overwhelming the node (ran trials on a test system before doing it in prod). I understand that just call counts gives less information, but I needed a very non-invasive method to start with.

josevalim commented:
> but I needed a very non-invasive method to start with.

if we follow this principle, then it feels the default should be call_count indeed. I will go ahead with that and finalize this PR.

josevalim commented:
I have concluded my changes. In summary:

* I have renamed `hprof` to `tprof` and it supports three measurements: `call_count`, `call_time`, and `call_memory`
* The default is `call_count`
* I have marked the current API as experimental in Erlang/OTP 27.0 - so we can fix any obvious mistake after release
* I have noted that `cprof` and `eprof` will be removed in Erlang/OTP 29.0

Once approved and merged, I will submit a PR to the compiler, so it supports profiling via tprof, allowing us to measure both time and heap allocations.

garazdawi commented on system/doc/general_info/upcoming_incompatibilities.xml:
```diff
@@ -331,6 +331,16 @@ String Content
+      <p>
+	As of OTP 29, the <c>cprof</c> and <c>eprof</c> will be
+	removed in favor of
+	<seemfa marker="tools:tprof"><c>tprof</c></seemfa> added
```
```suggestion
	<seeerl marker="tools:tprof"><c>tprof</c></seeerl> added
```

max-au commented:
The summary sounds good to me. I also agree with @garazdawi that `call_time` can be made the default (or, `call_count`, but it's less useful in most contexts).

FWIW, I never (ever) used `cprof`, and almost never used `eprof` (for it lacked features now available in `hprof` like profiling a process hierarchy). But I'm interested in hearing any examples of `cprof` usage.

jhogberg commented:
Merged, huge thanks for the PR! :)

dgud commented:
tprof:collect() is used in the example but not documented?

josevalim commented:
@dgud apparently I inherited that from `hprof` (it was not documented back then either). However, I forgot to update tprof:collect() results in the examples and that was my fault. If you do not have the bandwidth to fix it, please open up an issue and ping me, and I will do it. :)

dgud commented:
I wanted to test the memory tracer, which was great by the way, so I just followed the examples.
I guess the 'collect()' call is needed, I was mostly interested in if I need to call 'stop()' after 'collect()' or not.
I'll try to remember to take a look at it next week.

max-au commented:
> tprof:collect() is used in the example but not documented?

That's my oversight. And documentation linter did not point that out :-(
`collect/0` is used in many examples in the documentation, and is, of course, necessary.

There is no need to call `stop` after `collect` (and in fact you can collect as many times as you need).
